specs/validator/0_beacon-chain-validator.md

```diff
@@ -0,0 +1,383 @@
+)
+```
+
+#### Block body
+
+##### Proposer slashings
+
+Up to `MAX_PROPOSER_SLASHINGS` [`ProposerSlashing`](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#proposerslashing) objects can be included in the `block`. The proposer slashings must satisfy the verification conditions found in [proposer slashings processing](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#proposer-slashings-1). The validator receives a small "whistleblower" reward for each proposer slashing found and included.
+
+##### Casper slashings
+
+Up to `MAX_CASPER_SLASHINGS` [`CasperSlashing`](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#casperslashing) objects can be included in the `block`. The Casper slashings must satisfy the verification conditions found in [Casper slashings processing](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#casper-slashings-1). The validator receives a small "whistleblower" reward for each Casper slashing found and included.
+
+##### Attestations
+
+Up to `MAX_ATTESTATIONS` aggregate attestations can be included in the `block`. The attestations added must satisfy the verification conditions found in [attestation processing](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#attestations-1). To maximize profit, the validator should attempt to add aggregate attestations that include the most available that have not previously been added on chain.
+
+##### Deposits
+
+Up to `MAX_DEPOSITS` [`Deposit`](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#deposit) objects can be included in the `block`. These deposits are constructed from the `Deposit` logs from the [Eth1.0 deposit contract](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#ethereum-10-deposit-contract) and must be processed in sequential order. The deposits included in the `block` must satisfy the verification conditions found in [deposits processing](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#deposits-1).
+
+##### Exits
+
+Up to `MAX_EXITS` [`Exit`](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#exit) objects can be included in the `block`. The exits must satisfy the verification conditions found in [exits processing](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#exits-1).
+
+### Attestations
+
+A validator is expected to create, sign, and broadcast an attestation during each epoch. The slot during which the validator performs this role is any slot at which `get_shard_committees_at_slot(state, slot)` contains a committee that contains `validator_index`.
+
+A validator should create and broadcast the attestation halfway through the `slot` during which the validator is assigned -- that is `SLOT_DURATION * 0.5` seconds after the start of `slot`.
```

Comment:
>- Validators attesting to the expected beacon chain head during the previous epoch:
>    - Let `previous_epoch_head_attestations = [a for a in previous_epoch_attestations if a.data.beacon_block_root == get_block_root(state, a.data.slot)]`.
>    - Let `previous_epoch_head_attester_indices be the union of the validator index sets given by [get_attestation_participants(state, a.data, a.aggregation_bitfield) for a in previous_epoch_head_attestations]`.
>    - Let `previous_epoch_head_attesting_balance = sum([get_effective_balance(state, i) for i in previous_epoch_head_attester_indices])`.
> ...
> - Expected beacon chain head:
>    - Any validator `index` in `previous_epoch_head_attester_indices` gains `base_reward(state, index) * previous_epoch_head_attesting_balance // total_balance)`.
>    - Any active validator `index` not in `previous_epoch_head_attester_indices` loses `base_reward(state, index)`.

Doesn't this reward calculation logic make an intention to attest to a slot when block with attesting slot number is imported or, if block didn't show up, at the end of the slot? Does it make sense to attest to a slot right at the beginning of the next one?

Reply:
I'm not sure I understand the question. The intention of attesting to the block at the `slot` halfway through that `slot` is so that there is a chance for the most recent block to propagate to the attester and time for hte attester to broadcast the attestation to be included in the next block. (this is the ideal case at least)

Reply:
As I understand from reward calculation logic, attesting at the _end_ of the slot increases a chance to get reward with no negative side effects to validator. Cause it increases a chance to receive a block that eventually becomes a new head.

This condition `if a.data.beacon_block_root == get_block_root(state, a.data.slot)` says that reward will be gained only if attested block was a head of the chain at that particular slot. But the case is that head block could show up in _any_ moment during the slot, not only in its first half. And this incentivizes validator to defer attestation until the end of the slot, doesn't it?

Reply:
> to broadcast the attestation to be included in the next block. (this is the ideal case at least)

Including attestation right in the next block violates `MIN_ATTESTATION_INCLUSION_DELAY` condition.